endpoint/shellComp/shell.c

```diff
@@ -0,0 +1,496 @@
+  return sh_launch(args);
+}
+
+/**
+   @brief Builtin command: show the dns setting
+   @param[in] argv List of args
+   @return Always returns 1, to continue executing
+ */
+int sh_dns(char **argv) {
+  char *args[] = {"cat", "/etc/resolv.conf", NULL};
+  return sh_launch(args);
+}
+
+/**
+   @brief Builtin command: show the route setting
+   @param[in] argv List of args
+   @return Always returns 1, to continue executing
+ */
+int sh_route(char **argv) {
+  char *args[] = {"/sbin/route", "-n", NULL};
+  return sh_launch(args);
+}
+
+/**
+   @brief Builtin command: send the at command to AT device
+   @param[in] argv List of args
+   @return Always returns 1, to continue executing
+ */
+int sh_at(char **argv) {
+  if (argv[1] != NULL) {
+    size_t at_len = strlen(argv[1]);
+    char *at_cmd = malloc(at_len + 1);
+    memset(at_cmd, 0, at_len + 1);
+    memcpy(at_cmd, argv[1], at_len);
+    at_cmd[at_len] = '\n';
+    uart_write(at_fd, at_cmd);
+    char *msg = uart_read(at_fd);
+    uart_write(uart_fd, msg);
+    free(msg);
+    free(at_cmd);
+  }
+  return 1;
+}
+
+/**
+   @brief Send a transfer test to host
+   @param[in] host Host to send
+   @param[in] port Port to send
+   @return Always returns 1, to continue executing
+ */
+static int send_transaction_test(char *host, char *port) {
+  CURL *curl;
+  CURLcode res;
+
+  curl_global_init(CURL_GLOBAL_ALL);
+
+  /* get a curl handle */
+  curl = curl_easy_init();
+  if (curl) {
+    struct curl_slist *headers = NULL;
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+
+    size_t url_len = strlen(host) + strlen(port) + 21;
+
+    char *url = malloc(url_len);
+    if (url == NULL) {
+      LE_ERROR("Failed to allocated memory");
+      return 1;
+    }
+
+    snprintf(url, url_len, "http://%s:%s/transaction", host, port);
+
+    LE_INFO("Send transaction: %s", url);
+    if (url != NULL) {
+      curl_easy_setopt(curl, CURLOPT_URL, url);
+      curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, test_post_data);
+
+      /* Perform the request, res will get the return code */
+      res = curl_easy_perform(curl);
+      /* Check for errors */
+      if (res != CURLE_OK) fprintf(stderr, "curl failed: %s\n", curl_easy_strerror(res));
+      free(url);
+    }
+
+    /* always cleanup */
+    curl_easy_cleanup(curl);
+  }
+  curl_global_cleanup();
+  return 1;
+}
+
+/**
+   @brief Builtin command: send a test transfer
+   @param[in] argv List of args
+   @return Always returns 1, to continue executing
+ */
+int sh_transaction(char **argv) {
+  if (argv[1] == NULL || argv[2] == NULL) {
+    fprintf(stderr, "Should enter host and port\n");
+    return 1;
+  }
+
+  LE_INFO("Host: %s, Port: %s", argv[1], argv[2]);
+  return send_transaction_test(argv[1], argv[2]);
+}
+
+/**
+   @brief Builtin command: connect to LTE network
+   @param[in] argv List of args
+   @return Always returns 1, to continue executing
```

Comment:
Based on the description of `sh_launch`, it seems like it still has other possibility to return the value other than 1?

Reply:
The sh_exit() will return 0.